examples/I/hashmap.c

```diff
@@ -0,0 +1,132 @@
+};
+
+struct hashmap_keyval_t
+{
+    char* key;
+    value_t val;
+    struct hashmap_keyval_t* next;
+};
+
+
+static int hash_func(char* key)
+{
+	int len = strlen(key);
+    int hash = len;
+    int i;
+
+    for (i = 0; i < len; i++) {
+        hash ^= (int)key[i];
+    }
+
+    return hash;
+}
+
+struct hashmap_t* hashmap_create(int size, struct alloc_t alloc)
+{
+    int i;
+    
+    // use user provided allocator
+    struct hashmap_t* hashmap = alloc.malloc(sizeof(struct hashmap_t));
+    
+    hashmap->size = size;
+    hashmap->alloc = alloc;
+    hashmap->buckets = alloc.malloc(sizeof(struct hashmap_keyval_t*) * size);
+    for (i = 0; i < size; i++) {
+        hashmap->buckets[i] = NULL;
+    }
+
+    return hashmap;
+}
+
+void hashmap_destroy(struct hashmap_t* hashmap)
+{
+    if (hashmap == NULL) {
+        return;
+    }
+
+    int i;
+    for (i = 0; i < hashmap->size; i++) {
+        struct hashmap_keyval_t* iter = hashmap->buckets[i];
+        struct hashmap_keyval_t* next = hashmap->buckets[i]->next;
+        hashmap->alloc.free(iter);
+        while (next) {
+            iter = next;
+            next = iter->next;
+            hashmap->alloc.free(iter);            
+        }
+    }
+
+    hashmap->alloc.free(hashmap);
+}
+
+struct hashmap_keyval_t* hashmap_set(struct hashmap_t* hashmap, char* key, 
+    value_t val)
+{
+    if (hashmap == NULL) {
+        return NULL;
+    }
+
+    int hash = hash_func(key);
+    hash = hash % hashmap->size;
+    
+    // check if this is the first item in the bucket
+    if (hashmap->buckets[hash] == NULL) {
+        struct hashmap_keyval_t* keyval = 
+            hashmap->alloc.malloc(sizeof(struct hashmap_keyval_t));
+        keyval->key = key;
+        keyval->val = val;
+        keyval->next = NULL;
+        hashmap->buckets[hash] = keyval;
+        return keyval;
+    }
+
+    struct hashmap_keyval_t** iter = &hashmap->buckets[hash];
+    while (*iter) {
+        // check if the given key exists
+        if ((strlen((*iter)->key) == strlen(key)) && 
+            (strcmp((*iter)->key, key) == 0)) {
+            //update value
+            (*iter)->val = val;
+            return *iter;
+        }
+        iter = &(*iter)->next;
+    }
+
+    // add new value to the end of the bucket
+    *iter = hashmap->alloc.malloc(sizeof(struct hashmap_keyval_t));
+    (*iter)->key = key;
+    (*iter)->val = val;
+    (*iter)->next = NULL;
+    return *iter;
+}
+
+struct value_t* hashmap_get(struct hashmap_t* hashmap, char* key)
+{
+    int hash = hash_func(key);
+    hash = hash % hashmap->size;
+
+    if (hashmap->buckets[hash] == NULL) {
+        return NULL;
+    }
+
+    struct hashmap_keyval_t* iter = hashmap->buckets[hash];
+    while (iter) {
+        // check if the given key exists
+        if ((strlen(iter->key) == strlen(key)) && 
+            (strcmp(iter->key, key) == 0)) {
+            break;
+        }
+        iter = iter->next;
+    }
+
+    return iter->val;
```

Comment:
**Possible segmentation fault (nullptr dereference):**

Case: hash conflict && required key not on hashmap.

How: loop while (iter) ends by not finding particular key - it iterates over all existing keys and, on last iteration, assigns NULL to iter (line 128: iter = iter->next); while condition is not met and while loop is ended. Program reaches line 131 with iter == NULL.

Reply:
```suggestion
    return iter ? iter->val : NULL;
```

Reply:
fixed